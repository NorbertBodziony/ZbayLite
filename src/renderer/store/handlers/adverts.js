import BigNumber from 'bignumber.js'

import identitySelectors from '../selectors/identity'
import channelSelectors from '../selectors/channel'
import offersHandlers from '../../store/handlers/offers'
import { messageType } from '../../zbay/messages'
import { messages } from '../../zbay'
import { getClient } from '../../zcash'
import notificationsHandlers from './notifications'
import directMessagesQueueHandlers from './directMessagesQueue'
import { errorNotification, successNotification } from './utils'
import operationsHandlers, { operationTypes } from './operations'
import contactsHandlers from './contacts'

const handleSend = ({ values }) => async (dispatch, getState) => {
  const data = {
    tag: values.tag,
    background: values.background.toString(),
    title: values.title,
    provideShipping: values.shippingInfo ? '1' : '0',
    amount: values.usd,
    description: values.description
  }
  const identityAddress = identitySelectors.address(getState())
  const channel = channelSelectors.data(getState()).toJS()
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.AD,
      data: data,
      spent: new BigNumber('0.0001')
    },
    privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: channel.address,
    identityAddress
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(
      operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: {
          channelId: channel.id,
          message: message
        }
      })
    )
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification('Advert successfully posted')
      )
    )
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      dispatch(
        errorNotification({
          message: "Couldn't send the message, please check node connection."
        })
      )
    )
  }
}

const handleSendTransfer = ({ values, history, payload }) => async (dispatch, getState) => {
  const shippingData = identitySelectors.shippingData(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.ITEM_TRANSFER,
      data: {
        itemId: payload.id.substring(0, 64),
        tag: payload.tag,
        offerOwner: payload.offerOwner,
        shippingData
      },
      spent: new BigNumber(values.zec)
    },
    privKey
  })
  dispatch(offersHandlers.epics.createOfferAdvert({ payload, history }))
  dispatch(directMessagesQueueHandlers.epics.addDirectMessage({
    message,
    recipientAddress: payload.address,
    recipientUsername: payload.offerOwner
  })
  )
  dispatch(contactsHandlers.epics.updateDeletedChannelTimestamp({ address: payload.id + payload.offerOwner, timestamp: 0 }))
}

export const epics = {
  handleSend,
  handleSendTransfer
}

export default {
  epics
}
