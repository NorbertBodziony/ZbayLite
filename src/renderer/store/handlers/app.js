import { ipcRenderer, remote } from 'electron'
import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { actionTypes } from '../../../shared/static'
import { actionCreators } from './modals'

export const AppState = Immutable.Record(
  {
    version: null,
    transfers: Immutable.Map(),
    newUser: false,
    modalTabToOpen: null,
    allTransfersCount: 0,
    newTransfersCounter: 0
  },
  'AppState'
)

export const initialState = AppState()

const loadVersion = createAction(actionTypes.SET_APP_VERSION, () =>
  remote.app.getVersion()
)
const setTransfers = createAction(actionTypes.SET_TRANSFERS)
const setModalTab = createAction(actionTypes.SET_CURRENT_MODAL_TAB)
const clearModalTab = createAction(actionTypes.CLEAR_CURRENT_MODAL_TAB)
const setAllTransfersCount = createAction(actionTypes.SET_ALL_TRANSFERS_COUNT)
const setNewTransfersCount = createAction(actionTypes.SET_NEW_TRANSFERS_COUNT)
const reduceNewTransfersCount = createAction(
  actionTypes.REDUCE_NEW_TRANSFERS_COUNT
)

export const actions = {
  loadVersion,
  setTransfers,
  setModalTab,
  clearModalTab,
  setAllTransfersCount,
  setNewTransfersCount,
  reduceNewTransfersCount
}

export const askForBlockchainLocation = () => async (dispatch, getState) => {
  dispatch(actionCreators.openModal('blockchainLocation')())
}

export const proceedWithSyncing = (payload) => async (dispatch, getState) => {
  ipcRenderer.send('proceed-with-syncing', payload)
  dispatch(actionCreators.closeModal('blockchainLocation')())
}

export const reducer = handleActions(
  {
    [setNewTransfersCount]: (state, { payload: setNewTransfersCount }) =>
      state.set('newTransfersCounter', setNewTransfersCount),
    [reduceNewTransfersCount]: (state, { payload: amount }) =>
      state.update('newTransfersCounter', count => count - amount),
    [loadVersion]: (state, { payload: version }) =>
      state.set('version', version),
    [setModalTab]: (state, { payload: tabName }) =>
      state.set('modalTabToOpen', tabName),
    [setAllTransfersCount]: (state, { payload: transfersCount }) =>
      state.set('allTransfersCount', transfersCount),
    [clearModalTab]: state => state.set('modalTabToOpen', null),
    [setTransfers]: (state, { payload: { id, value } }) => {
      return state.setIn(['transfers', id], value)
    }
  },
  initialState
)

export const epics = {
  askForBlockchainLocation,
  proceedWithSyncing
}

export default {
  epics,
  actions,
  reducer
}
