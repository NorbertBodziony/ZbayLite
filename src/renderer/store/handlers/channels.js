import Immutable from 'immutable'
import { DateTime } from 'luxon'
import { createAction, handleActions } from 'redux-actions'
import BigNumber from 'bignumber.js'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification,
  LoaderState
} from './utils'
import notificationsHandlers from './notifications'
import channelHandlers from './channel'
import identitySelectors from '../selectors/identity'
import channelsSelectors from '../selectors/channels'
import modalsHandlers from './modals'
import { messages } from '../../zbay'
import { getVault } from '../../vault'
import { getClient } from '../../zcash'
import { networkFee } from '../../../shared/static'

const toBigNumber = x => new BigNumber(x)

export const ChannelsState = Immutable.Record(
  {
    data: Immutable.List(),
    loader: LoaderState({ loading: true })
  },
  'ChannelsState'
)

export const initialState = ChannelsState()

export const actionTypes = {
  LOAD_CHANNELS: 'LOAD_IDENTITY_CHANNELS',
  SET_LAST_SEEN: 'SET_CHANNELS_LAST_SEEN',
  SET_UNREAD: 'SET_CHANNEL_UNREAD'
}

const loadChannels = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)

  return channels
})
const loadChannelsToNode = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)
  await Promise.all(
    channels.map(channel =>
      getClient().keys.importIVK({ ivk: channel.keys.ivk, address: channel.address })
    )
  )
  await Promise.all(
    channels.filter(ch => ch.keys.sk).map(channel =>
      getClient().keys.importSK({ sk: channel.keys.sk, rescan: 'no' })
    )
  )
  return channels
})

const setLastSeen = createAction(actionTypes.SET_LAST_SEEN)
const setUnread = createAction(actionTypes.SET_UNREAD)

export const actions = {
  loadChannels,
  setLastSeen,
  setUnread,
  loadChannelsToNode
}

const _createChannel = async (identityId, { name, description }) => {
  const address = await getClient().addresses.create('sapling')
  const [ivk, sk] = await Promise.all([
    getClient().keys.exportIVK(address),
    getClient().keys.exportSK(address)
  ])
  const channel = {
    address,
    name: name,
    description: description,
    private: true,
    keys: {
      ivk,
      sk
    }
  }
  await getVault().channels.importChannel(identityId, channel)
  return address
}

const createChannel = (values, formActions) => async (dispatch, getState) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  const balance = identitySelectors.balance('zec')(getState())
  try {
    if (balance.lt(0.0002)) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({ message: `You need minimum 0.0002 ZEC to create a channel. ` })
        )
      )
      formActions.setSubmitting(false)
      return
    }
    const identityId = identitySelectors.id(getState())
    const address = await _createChannel(identityId, values)
    await dispatch(channelHandlers.epics.sendChannelSettingsMessage({ address: address }))
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({ message: `Successfully created ${values.name} channel.` })
      )
    )
    formActions.setSubmitting(false)
    dispatch(closeModal())
    dispatch(loadChannels(identityId))
  } catch (error) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({ message: `Failed to create channel: ${error.message}` })
      )
    )
    formActions.setSubmitting(false)
  }
}

const withdrawMoneyFromChannels = () => async (dispatch, getState) => {
  const ownedChannels = channelsSelectors.ownedChannels(getState())
  let earnedAmount = toBigNumber(0)
  for (const channel of ownedChannels) {
    const channelAddress = channel.get('address')
    const amount = await getMoneyFromChannel(channelAddress)(dispatch, getState)
    earnedAmount = earnedAmount.plus(amount)
  }
  if (earnedAmount.gt(toBigNumber(0))) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `You will shortly receive ${earnedAmount.toFixed(
            4
          )} zash for commission from your created channels.`
        })
      )
    )
  }
}

const getMoneyFromChannel = address => async (dispatch, getState) => {
  const amount = await getClient().accounting.balance(address)
  const identityAddress = identitySelectors.address(getState())
  if (amount.gt(0.0005)) {
    const transfer = messages.createEmptyTransfer({
      address: identityAddress,
      amount: amount.minus(networkFee).toString(),
      identityAddress: address
    })
    await getClient().payment.send(transfer)
    return amount.minus(networkFee)
  }
  return toBigNumber(0)
}

const updateLastSeen = ({ channelId }) => async (dispatch, getState) => {
  const identity = identitySelectors.data(getState())
  const lastSeen = DateTime.utc()
  await getVault().channels.updateLastSeen({
    identityId: identity.get('id'),
    channelId,
    lastSeen
  })
  dispatch(setLastSeen({ channelId, lastSeen }))
  dispatch(setUnread({ channelId, unread: 0 }))
}

export const epics = {
  createChannel,
  updateLastSeen,
  getMoneyFromChannel,
  withdrawMoneyFromChannels
}

export const reducer = handleActions(
  {
    [typePending(actionTypes.LOAD_CHANNELS)]: state =>
      state.setIn(['loader', 'loading'], true).setIn(['loader', 'message'], 'Loading channels'),
    [typeFulfilled(actionTypes.LOAD_CHANNELS)]: (state, { payload: data }) =>
      state.set('data', Immutable.fromJS(data)).setIn(['loader', 'loading'], false),
    [typeRejected(actionTypes.LOAD_CHANNELS)]: (state, { payload: error }) =>
      state.setIn(['loader', 'loading'], false),
    [setLastSeen]: (state, { payload: { channelId, lastSeen } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('lastSeen', lastSeen))
    },
    [setUnread]: (state, { payload: { channelId, unread } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('unread', unread))
    }
  },
  initialState
)

export default {
  actions,
  epics,
  reducer
}
