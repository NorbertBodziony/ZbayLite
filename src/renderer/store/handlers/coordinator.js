import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import channelsSelectors from '../selectors/channels'
import coordinatorSelectors from '../selectors/coordinator'
import messagesHandlers from './messages'
import contactsHandlers from './contacts'
import nodeHandlers from './node'
import identityHandlers from './identity'
import usersHandlers from './users'

export const Coordinator = Immutable.Record(
  {
    running: true
  },
  'Coordinator'
)

export const initialState = Coordinator()

export const stopCoordinator = createAction('STOP_COORDINATOR')
export const startCoordinator = createAction('START_COORDINATOR')

const actions = {
  stopCoordinator,
  startCoordinator
}
const coordinator = () => async (dispatch, getState) => {
  let index = 0
  while (coordinatorSelectors.running(getState())) {
    const actions = channelsSelectors
      .data(getState())
      .map(channel => () => messagesHandlers.epics.fetchMessages(channel))
      .push(() => contactsHandlers.epics.fetchMessages())
      .push(() => nodeHandlers.epics.getStatus())
      .push(() => identityHandlers.epics.fetchBalance())
      .push(() => usersHandlers.epics.fetchUsers())
    await dispatch(actions.get(index % actions.size)())
    index += 1
  }
}
const epics = {
  coordinator
}

export const reducer = handleActions(
  {
    [startCoordinator]: state => state.set('running', true),
    [stopCoordinator]: state => state.set('running', false)
  },
  initialState
)

export default {
  epics,
  actions,
  reducer
}
