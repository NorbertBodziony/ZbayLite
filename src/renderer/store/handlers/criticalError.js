import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { actionTypes } from '../../../shared/static'

export const CriticalError = Immutable.Record({
  message: '',
  traceback: ''
}, 'CriticalError')

export const initialState = CriticalError()

const setCriticalError = createAction(actionTypes.SET_CRITICAL_ERROR)

export const actions = {
  setCriticalError
}

export const reducer = handleActions({
  [setCriticalError]: (state, { payload: error }) => state.merge(error)
}, initialState)

export default {
  actions,
  reducer
}
