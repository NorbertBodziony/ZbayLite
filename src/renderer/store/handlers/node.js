import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'

import { typeRejected, LoaderState } from './utils'
import { getClient } from '../../zcash'

const DEFAULT_ADDRESS_TYPE = 'sapling'

export const NodeState = Immutable.Record({
  latestBlock: new BigNumber(0),
  currentBlock: new BigNumber(0),
  connections: new BigNumber(0),
  isTestnet: null,
  status: 'connecting',
  errors: '',
  bootstrapLoader: LoaderState(),
  startedAt: null
}, 'NodeState')

export const initialState = NodeState()

export const actionTypes = {
  GET_STATUS: 'GET_NODE_STATUS',
  CREATE_ADDRESS: 'CREATE_ZCASH_ADDRESS',
  GET_BALANCE: 'GET_BALANCE',
  SET_STATUS: 'SET_NODE_STATUS',
  SET_BOOTSTRAPPING: 'SET_NODE_BOOTSTRAPPING',
  SET_BOOTSTRAPPING_MESSAGE: 'SET_NODE_BOOTSTRAPPING_MESSAGE'
}

const setStatus = createAction(actionTypes.SET_STATUS)

const createAddress = createAction(
  actionTypes.CREATE_ADDRESS,
  async (type = DEFAULT_ADDRESS_TYPE) => {
    return getClient().addresses.create(type)
  }
)

const setBootstrapping = createAction(actionTypes.SET_BOOTSTRAPPING)
const setBootstrappingMessage = createAction(actionTypes.SET_BOOTSTRAPPING_MESSAGE)

const actions = {
  createAddress,
  setBootstrapping,
  setBootstrappingMessage
}

const getStatus = () => async (dispatch) => {
  try {
    const info = await getClient().status.info()
    dispatch(setStatus(info))
  } catch (err) {
    dispatch(setStatus({ 'status': 'down', errors: err }))
  }
}

const restart = () => (dispatch) => {
  console.log('Restarting node')
}

const togglePower = () => (dispatch) => {
  console.log('toggling power of the node')
}

const epics = {
  getStatus,
  restart,
  togglePower
}

export const reducer = handleActions({
  [actionTypes.SET_STATUS]: (state, { payload: status }) => state.merge(status),
  [typeRejected(actionTypes.CREATE_ADDRESS)]: (state, { payload: errors }) => state.set('errors', errors),
  [setBootstrapping]: (state, { payload: bootstrapping }) => state.setIn(['bootstrapLoader', 'loading'], bootstrapping),
  [setBootstrappingMessage]: (state, { payload: message }) => state.setIn(['bootstrapLoader', 'message'], message)
}, initialState)

export default {
  actions,
  reducer,
  epics
}
