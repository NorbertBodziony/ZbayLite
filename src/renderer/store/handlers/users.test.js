/* eslint import/first: 0 */
jest.mock('../../zcash')
jest.mock('../../vault')

import Immutable from 'immutable'
import * as R from 'ramda'

import handlers, { initialState, ReceivedUser } from './users'
import { NodeState } from './node'
import selectors from '../selectors/users'
import { packMemo } from '../../zbay/transit'
import { mock as zcashMock } from '../../zcash'
import { ChannelsState } from './channels'

import create from '../create'
import testUtils from '../../testUtils'

describe('users reducer', () => {
  let store = null

  const userChannel = Immutable.fromJS([testUtils.channels.createUsersChannel(1)])

  beforeEach(() => {
    store = create({
      initialState: Immutable.Map({
        node: NodeState({ isTestnet: true }),
        users: initialState,
        channels: ChannelsState({
          data: userChannel
        })
      })
    })
    jest.clearAllMocks()
  })

  describe('handles actions -', () => {
    describe('setUsers', () => {
      it('sets users when empty', () => {
        const users = R.range(0, 3)
          .map(id => testUtils.messages.createReceivedUserMessage({ id: `test-user-${id}` }))
          .reduce(
            async (acc = Promise.resolve(Immutable.Map({})), message) => {
              const accumulator = await acc
              return Promise.resolve(
                accumulator.merge(ReceivedUser(message, 0)).merge(ReceivedUser(message, 1))
              )
            },
            Promise.resolve(Immutable.Map({}))
          )
        const users2 = R.range(3, 5)
          .map(id => testUtils.messages.createReceivedUserMessage({ id: `test-user-${id}` }))
          .reduce(
            async (acc = Promise.resolve(Immutable.Map({})), message) => {
              const accumulator = await acc
              return Promise.resolve(
                accumulator.merge(ReceivedUser(message, 0)).merge(ReceivedUser(message, 1))
              )
            },
            Promise.resolve(Immutable.Map({}))
          )

        store.dispatch(handlers.actions.setUsers({ users }))
        store.dispatch(handlers.actions.setUsers({ messages: users2 }))

        expect(selectors.users(store.getState())).toMatchSnapshot()
      })
    })
  })

  describe('handles epics -', () => {
    describe('fetch users', () => {
      const _createMessagesForUsers = num => async address =>
        Promise.all(
          R.range(0, num).map(async i => {
            const message = testUtils.messages.createSendableUserMessage({
              message: {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'nickname',
                address:
                  'ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r'
              },
              createdAt: testUtils.now.minus({ hours: i }).toSeconds()
            })
            return testUtils.transfers.createTransfer({
              txid: `tx-id-${i}-${address}`,
              memo: await packMemo(message)
            })
          })
        )

      it('when users', async () => {
        await zcashMock.requestManager.z_listreceivedbyaddress.mockImplementation(
          _createMessagesForUsers(2)
        )
        await store.dispatch(handlers.epics.fetchUsers())
        expect(selectors.users(store.getState())).toMatchSnapshot()
      })
    })

    describe('isNicknameTaken', () => {
      beforeEach(() => {
        store = create({
          initialState: Immutable.Map({
            node: NodeState({ isTestnet: true }),
            users: Immutable.fromJS({
              '02c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2392': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-1',
                address: 'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhyb4c'
              },
              '02c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c23vb': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-1',
                address: 'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhyb4c'
              },
              '04c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2355': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-2',
                address: 'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhylnn'
              },
              '04c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2vbf': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-2',
                address: 'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhylnn'
              }
            })
          })
        })
        jest.clearAllMocks()
      })

      it('should return true for taken username', async () => {
        const isNicknameTaken = await store.dispatch(handlers.epics.isNicknameTaken('test-user-2'))
        expect(isNicknameTaken).toBeTruthy()
      })

      it('should return false for available username', async () => {
        const isNicknameTaken = await store.dispatch(handlers.epics.isNicknameTaken('test-new-user'))
        expect(isNicknameTaken).toBeFalsy()
      })
    })

    describe('isNicknameTaken on empty store', () => {
      beforeEach(() => {
        store = create({
          initialState: Immutable.Map({
            node: NodeState({ isTestnet: true }),
            users: initialState
          })
        })
        jest.clearAllMocks()
      })

      it('should return false for available username when no users', async () => {
        const isNicknameTaken = await store.dispatch(handlers.epics.isNicknameTaken('test-user-2'))
        expect(isNicknameTaken).toBeFalsy()
      })
    })
  })
})
