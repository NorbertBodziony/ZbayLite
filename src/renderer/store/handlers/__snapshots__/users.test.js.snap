// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users reducer handles actions - setUsers sets users when empty 1`] = `
Immutable.Map {
  "039ea1be861812ac6e210c96ac54fd2523f6d738a2db460613cc9aa8bd1886607f": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "034b0c8c8555259d37b4193e03a09d5f803ebbcb1988edc72fe3d6912a4687512d": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "0214dc56ab3467e74b3bc1fed80d08e0146b251caceddc2ad6bfe37caca5fcc2e5": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #2",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "0279b8ef97aa2c9d402c7c86dcc37e3f979165450ac2bc708bc6897ae33af97d52": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #2",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "026918cbe1c9ace24ff065006c06b3c6344a5d7203ad42a45ddc7c7a41c6edfcc2": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #3",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "03dfe22ffb15850264c08462781e228682e9d333da50d395ab07f2a1535b595444": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #3",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
}
`;

exports[`users reducer handles actions - setUsers sets users with same usernames 1`] = `
Immutable.Map {
  "039ea1be861812ac6e210c96ac54fd2523f6d738a2db460613cc9aa8bd1886607f": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #6",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "034b0c8c8555259d37b4193e03a09d5f803ebbcb1988edc72fe3d6912a4687512d": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #6",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "0214dc56ab3467e74b3bc1fed80d08e0146b251caceddc2ad6bfe37caca5fcc2e5": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #7",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "0279b8ef97aa2c9d402c7c86dcc37e3f979165450ac2bc708bc6897ae33af97d52": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #7",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "026918cbe1c9ace24ff065006c06b3c6344a5d7203ad42a45ddc7c7a41c6edfcc2": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #8",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "03dfe22ffb15850264c08462781e228682e9d333da50d395ab07f2a1535b595444": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #8",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
}
`;

exports[`users reducer handles epics - createOrUpdateUser sends registration message 1`] = `
Array [
  Array [
    "ztestsapling1mwmpvwy2aah7dlt0c9l47gde982xv6snxq2srddfzx8efn2qht6mxnz65rtst0426gtxje0eqlm",
    Array [
      Object {
        "address": "ztestsapling1ggtntm4dgfc0remdxzex0pz722swjtqxc7wxj4eeqksmups04dkhjvcj5vcklmud2y3yvgj302m",
        "amount": "0.0001",
        "memo": "051778bce918ceca3d656a49c73f316b86e8bab1f3c9f9ee30214ac4c0581b78bc1f5773823720a125dea71e3d71274035769b2e5ff3f74d3dfc5033a2ba58d1955c8116b4746573746e616d65000000000000000000000000746573746c6173746e616d6500000000000000006e69636b6e616d65000000000000000000000000027a746573747361706c696e67316d776d707677793261616837646c743063396c3437676465393832787636736e78713273726464667a783865666e32716874366d786e7a363572747374303432366774786a653065716c6d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      },
    ],
    1,
    0.0001,
  ],
]
`;

exports[`users reducer handles epics - fetch users when users 1`] = `
Immutable.Map {
  "035302fc10b2dc5cbe747779cd1cbcd7fe111161e09282e210d72dee9f6e68b17a": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #9",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "037128c776aa4b67f744425a67968562667d8793c4ea52aef214a912772f80ebf0": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #9",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "02b7e647acb166a81139b1bcbdde82b8df610f1fb93af5edd0359dd0d7138d9166": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #10",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
  "03c5eb408fd7016d9f27547abd54110de58644e4d29b63cee5401a51e1f409df67": Immutable.UserData {
    "firstName": "testname",
    "lastName": "testlastname",
    "nickname": "nickname #10",
    "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
  },
}
`;
